ca_gazebo/src/create_virtual_wall_plugin.cpp

```diff
@@ -0,0 +1,147 @@
+  // simulation iteration.
+  this->updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+      std::bind(&VirtualWallSensorPlugin::OnUpdate, this));
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "Loaded");
+}
+
+void VirtualWallSensorPlugin::OnAddEntity()
+{
+  const auto & world_models = this->world_->Models();
+  // Filter models which name starts with `irobot_create2.` and are continued with a number.
+  create_2_models_.resize(0);
+  pubs_.resize(0);
+  std::copy_if(
+    world_models.begin(),
+    world_models.end(),
+    std::back_inserter(this->create_2_models_),
+    [this] (const physics::ModelPtr & model)
+    {
+      const std::string & name = model->GetName();
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "checking model %s", name.c_str());
+      if (name.size() < create2_model_name_prefix_length + 1) {
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name is short");
+        return false;
+      }
+      if (std::strncmp(
+        name.c_str(),
+        create2_model_name_prefix,
+        create2_model_name_prefix_length) != 0)
+      {
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name doesn't start with prefix");
+        return false;
+      }
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "%s", name.substr(create2_model_name_prefix_length).c_str());
+      std::istringstream iss(name.substr(create2_model_name_prefix_length));
+      size_t x;
+      iss >> x;
+      if (iss) {
+        std::ostringstream oss;
+        oss << "create" << x << "/virtual_wall";
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name ok");
+        this->pubs_.emplace_back(
+          this->rosnode_->advertise<std_msgs::Bool>(
+            oss.str().c_str(), 1));
+        return true;
+      }
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "model name doesn't end with a number");
+      return false;
+    });
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "number of create2: %zu", create_2_models_.size());
+}
+
+void VirtualWallSensorPlugin::OnUpdate()
+{
+  double detected_range = this->sensor_->Range(0);
+  ignition::math::Vector3d laser_dir(1., 0., 0.);
+  ignition::math::Pose3d sensor_pose(-0.145, 0., 0.068, 0., 0., 0.);
+  // 0.145, 0.308
+  const auto & parent_pose = world_->EntityByName(sensor_->ParentName())->WorldPose();
+  const ignition::math::Pose3d & laser_pose = parent_pose + sensor_->Pose();
+
+  laser_dir = laser_pose.Rot().RotateVector(laser_dir);
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "pose x=%f y=%f z=%f",
+    laser_pose.Pos().X(),
+    laser_pose.Pos().Y(),
+    laser_pose.Pos().Z());
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "dir x=%f y=%f z=%f",
+    laser_dir.X(),
+    laser_dir.Y(),
+    laser_dir.Z());
+
+  for (size_t i = 0; i < create_2_models_.size(); i++) {
+    const auto & model = this->create_2_models_[i];
+    const auto & pub = this->pubs_[i];
+
+    const ignition::math::Pose3d & robot_pose = model->WorldPose();
+    sensor_pose = robot_pose + sensor_pose;
```

Comment:
Does this work?

```c++
sensor_pose += robot_pose;
```

Reply:
`robot_pose + sensor_pose` is not the same as `sensor_pose + robot_pose` (pose addition isn't commutative).